packages/core/src/input/pointer/PointerManager.ts

```diff
@@ -81,32 +68,40 @@ export class PointerManager implements IInput {
    * @internal
    */
   _update(frameCount: number): void {
-    this._needOverallPointers && this._overallPointers();
-    this._downList.length = 0;
-    this._upList.length = 0;
-    this._movingDelta.set(0, 0);
-    this._nativeEvents.length > 0 && this._handlePointerEvent(this._nativeEvents, frameCount);
-    if (this._engine.physicsManager._initialized) {
-      const rayCastEntity = this._pointerRayCast();
-      const { _keyEventCount: keyEventCount } = this;
-      if (keyEventCount > 0) {
-        const { _keyEventList: keyEventList } = this;
-        for (let i = 0; i < keyEventCount; i++) {
-          switch (keyEventList[i]) {
-            case PointerKeyEvent.Down:
-              this._firePointerDown(rayCastEntity);
-              break;
-            case PointerKeyEvent.Up:
-              this._firePointerUpAndClick(rayCastEntity);
-              break;
-          }
+    const { _pointers: pointers, _nativeEvents: nativeEvents } = this;
+    /** Clean up the pointer released in the previous frame. */
+    let length = pointers.length;
+    if (length > 0) {
+      for (let i = length - 1; i >= 0; i--) {
+        if (i !== length - 1) {
+          pointers[i] = pointers[length - 1];
+          --length;
         }
-        this._firePointerExitAndEnter(rayCastEntity);
-        keyEventList[keyEventCount - 1] === PointerKeyEvent.Leave && (this._currentPressedEntity = null);
-        this._keyEventCount = 0;
-      } else {
-        this._firePointerDrag();
-        this._firePointerExitAndEnter(rayCastEntity);
+      }
+      pointers.length = length;
+    }
+
+    /** Generate the pointer received for this frame. */
+    length = nativeEvents.length;
+    if (length > 0) {
+      this._buttons = nativeEvents[length - 1].buttons;
+      for (let i = 0; i < length; i++) {
+        const evt = nativeEvents[i];
+        this._getPointer(evt.pointerId, frameCount)?._events.push(evt);
+      }
+      nativeEvents.length = 0;
+    }
+
+    /** Pointer handles its own events. */
+    length = pointers.length;
+    if (length > 0) {
+      const updatePointer = this._engine.physicsManager._initialized
+        ? this._updatePointerWithPhysics
+        : this._updatePointerWithoutPhysics;
+      const { clientWidth, clientHeight } = this._htmlCanvas;
```

Comment:
Can merge  `this._updatePointerWithPhysics` and `this._updatePointerWithoutPhysics` logic?